Thread synchronization in concurrence

cburbridge opened the pull request:
Adding event for thread synchronization in concurrence and using event not condition in monitor state.  https://github.com/ros/executive_smach/issues/25

cburbridge committed:
Adding event for thread synchronization in concurrence and using event not condition in monitor state

cburbridge commented:
I think this solves the issue we had this morning.

hawesie commented:
Should we revert the fix from this morning? Have you pull-requested this upstream too?

cburbridge committed:
if monitor state prempted before executing, return.

cburbridge commented:
I have not made an upstream yet.

cburbridge committed:
cope with missed state termination notifications

Concurrent states could terminate and notify _ready_event without the concurrence container realising, as it could be busy checking the outcome values. This makes the concurrency container get stuck on line 250. This commit adds a timeout to the wait to safely cope with missing notifications.

jbohren commented on smach/src/smach/concurrence.py:
```diff
@@ -238,7 +247,7 @@ def execute(self, parent_ud = smach.UserData()):
                 break
             self._done_cond.acquire()
-            self._done_cond.wait()
+            self._done_cond.wait(0.1)
```
Is this magic timeout in the `done` condition check necessary?

cburbridge replied:
I think it is, because otherwise it is possible for the program to block waiting for a thread to notify that it is done when there are no more threads running. For example, while the main thread is in line 247, if a state runner thread notifies, then it will be missed. If this is the last running thread, then it causes a infinite wait.
An alternative might be to move the `_done_cond.aquire()` to the to before the outcome check, although I think then it is still possible to miss a thread notification when checking `not smach.is_shutdown()`

cburbridge committed:
Checking threads have fully terminated before cleanup of outcomes dict

This commit uses thread.isAlive() on each concurrent state runner to check for termination of all the threads before continuing. This is necessary as only checking that the outcome has been filled in does not mean the thread has completed; if the thread has not completed it may not yet have called the termination callback. If this loop exits before the termination callback of the last thread is called, then the callback will occasionally be sent an empty dictionary (when the main thread has got to line 305).

cburbridge commented:
This last commit fixes the first bug reported on https://github.com/ros/executive_smach/issues/25
